moment_kinetics/src/fokker_planck_calculus.jl

```diff
@@ -2317,4 +2318,97 @@ function enforce_vpavperp_BCs!(pdf,vpa,vperp,vpa_spectral,vperp_spectral)
     end
 end
 
+"""
+function to interpolate f(vpa,vperp) from one 
+velocity grid to another, assuming that both 
+grids are represented by vpa, vperp in normalised units,
+but have different normalisation factors 
+defining the meaning of these grids in physical units.
+
+E.g. vpai, vperpi = ci * vpa, ci * vperp
+     vpae, vperpe = ce * vpa, ce * vperp
+     
+with ci = sqrt(Ti/mi), ce = sqrt(Te/mi)
+
+scalefac = ci / ce is the ratio of the
+two reference speeds
+
+"""
+function interpolate_2D_vspace!(pdf_out,pdf_in,vpa,vperp,scalefac)
+    
+    begin_anyv_vperp_vpa_region()
+    # loop over points in the output interpolated dataset
+    @loop_vperp ivperp begin
+        vperp_val = vperp.grid[ivperp]*scalefac
+        # get element for interpolation data
+        iel_vperp = ielement_loopup(vperp_val,vperp)
+        if iel_vperp < 1 # vperp_interp outside of range of vperp.grid
+            @loop_vpa ivpa begin
+                pdf_out[ivpa,ivperp] = 0.0
+            end
+            continue
+        else
+            # get nodes for interpolation
+            ivperpmin, ivperpmax = vperp.igrid_full[1,iel_vperp], vperp.igrid_full[vperp.ngrid,iel_vperp]
+            vperp_nodes = vperp.grid[ivperpmin:ivperpmax]
+            #print("vperp: ",iel_vperp, " ", vperp_nodes," ",vperp_val)
+                   
+        end
+        @loop_vpa ivpa begin
+            vpa_val = vpa.grid[ivpa]*scalefac
+            # get element for interpolation data
+            iel_vpa = ielement_loopup(vpa_val,vpa)
+            if iel_vpa < 1 # vpa_interp outside of range of vpa.grid
+                pdf_out[ivpa,ivperp] = 0.0
+                continue
+            else
+                # get nodes for interpolation
+                ivpamin, ivpamax = vpa.igrid_full[1,iel_vpa], vpa.igrid_full[vpa.ngrid,iel_vpa]
+                vpa_nodes = vpa.grid[ivpamin:ivpamax]
+                #print("vpa: ", iel_vpa, " ", vpa_nodes," ",vpa_val)
+                   
+                # do the interpolation
+                pdf_out[ivpa,ivperp] = 0.0
+                for ivperpgrid in 1:vperp.ngrid
+                   # index for referencing pdf_in on orginal grid
+                   ivperpp = vperp.igrid_full[ivperpgrid,iel_vperp]
+                   # interpolating polynomial value at ivperpp for interpolation
+                   vperppoly = lagrange_poly(ivperpgrid,vperp_nodes,vperp_val)
+                   for ivpagrid in 1:vpa.ngrid
+                       # index for referencing pdf_in on orginal grid
+                       ivpap = vpa.igrid_full[ivpagrid,iel_vpa]
+                       # interpolating polynomial value at ivpap for interpolation
+                       vpapoly = lagrange_poly(ivpagrid,vpa_nodes,vpa_val)
+                       pdf_out[ivpa,ivperp] += vpapoly*vperppoly*pdf_in[ivpap,ivperpp]
+                   end
+                end
+            end
+        end
+    end
+    return nothing
+end

```

Comment:
If this interpolation is a significant part of the run-time - so worth optimising - it might be worth thinking about the following.
I think (although I haven't checked carefully so there could well be mistakes in the following!) two 1D interpolations should be significantly cheaper than a 2D interpolation, and is probably exactly equivalent (apart from round-off errors). The cost of one interpolation is something like `nout * npoly` where `nout` is the number of grid points of the output, and `npoly` is the number of coefficients in the interpolating polynomial (from the input) that have to be evaluated.
So the cost for two 1D interpolations is:
```
nvperp * (interpolation in vpa) + nvpa * (interpolation in vperp)
= nvperp * nvpa * vpa_ngrid + nvpa * nvperp * vperp_ngrid
```
while the cost for the 2D interpolation is
```
nvperp * nvpa * (vpa_ngrid * vperp_ngrid)
```
so the cost is larger than two 1D interpolations by `vpa_ngrid * vperp_ngrid / (vpa_ngrid + vperp_ngrid)`, which for `vpa_ngrid=vperp_ngrid=5` would be a factor of 2.5, or 4.5 for `vpa_ngrid=vperp_ngrid=9`.


Doing the operation in terms of 1D interpolation would also make me happy because it would mean that you could implement `interpolate_to_grid_1d!` for `gauss_legendre` and it would simplify this function (although you do need a funny-sized intermediate buffer array which might be a bit of a pain...) and also enable using `gauss_legendre` with the interpolation everywhere else in moment_kinetics :smiley:

Reply:
Currently the 2D interpolation is not used for any run-time process. This functionality was added to prepare for the Rosenbluth potential solve for then s' species are arbitrary distributions. However, I did not finish that feature yet because I did not have a definite normalisation scheme for the multi-species moment kinetics. If you can provide a definite set of normalisations, I can finish this feature.

I do not think that 2 1D interpolations is equivalent to a single 2D interpolation unless the function is exactly represented by a polynomial (and the sampling is of the right order). The error will be of order of the discretisation error. Perhaps you imagine that the operations are all machine precision accurate? I do not think that is realistic for the functions that we are trying to solve for here, with the affordable resolutions available.

The reason that I went for the 2D interpolation is that we are using the outer product representation
```math
f(x,y) = \sum_{i,j} \hat{f}_{ij}\phi_i(x) \phi_j(y), \quad \hat{f}_{ij} = f(x_i,y_j)
```
where the 2D interpolation is obtained by sending $x\rightarrow x_m$ and $y\rightarrow y_n$ 
If we use 2 1D interpolations, I think you are arguing for first obtaining the set of points  $x = x_m \neq x_i$ and $y = y_j$, then for every $x_m$, you use the data at $(x_m,y_j)$ to find the data at $(x_m,y_n\neq y_j)$.

Reply:
@mrhardman, yes what you describe is what I was suggesting.

The '2D interpolation' is, like you said, evaluating the 2D polynomial interpolating function, defined on the original grid, at some new points. In particular at $x=x_m$ the 2D interpolating polynomial (with degree $(n_x -1) \times (n_y - 1)$ ) becomes a 1D polynomial $f(x_m,y)$, and its coefficients are uniquely specified by values at $n_y$ points, i.e. $(x_m, y_j)$ for each $j$. So the second 1D interpolation should, I think, be evaluation of a 1D slice of the original 2D interpolating polynomial, so that the combined effect of the two 1D interpolations is equivalent (apart from rounding errors) to the 2D interpolation.

Anyway if the interpolating is only done during initialisation, this is not important. I guess I imagine that eventually, at least for a moment-kinetic case, we would need to interpolate between the grids for different species at each time step, and then the performance of the interpolation might become more of an issue.

Reply:
It sounds like there is a premise to test there -- we can do the check on the two different methods in a further PR (or reopen this branch & PR) and see what the accuracy and performance is like. For now this interpolation routine is only used in the testing module. The interpolation routine in the runtime collision operator would have to be used every time step regardless of whether we were moment kinetic or not (we have to keep moving Fi -> the grid of Fe and vice versa).